NUT-19: Cached Responses (#195)
cashubtc/nuts: XX.md

NUT-XX: Cached Responses
==========================

`optional`

---

To mitigate the risk of potential ecash loss due to network errors during critical operations such as minting, swapping, and melting, we introduce a caching mechanism for successful responses. This mechanism uses the `POST` request path and body contents to derive a key and then cache the responses using that key. If an identical request is received, the system will provide the cached response, minimizing the risk of ecash loss in the event of network disruptions.
 

## Requests & Responses

> Any Mint implementation should elect a data structure `D` that maps request objects to their respective responses. `D` should be fit for fast insertion, look-up and deletion (eviction) operations.

Reviewer: Example: Redis?

Author: ?

### Derive & Repeat

> Upon the reception of a mint (`POST /v1/mint/{method}`), swap (`POST v1/swap`) or melt (`POST /v1/melt/{method}`) `request` Bob should immediately derive a key `k` for it. `k` should depend on the path of `request` as well as the contents of `request`'s payload.

Reviewer: > reception of a mint (`POST /v1/mint/{method}`), swap (`POST v1/swap`) or melt (`POST /v1/melt/{method}`) `request`

Should refer and link to NUTs

> Upon the reception of a mint (`POST /v1/mint/{method}`), swap (`POST v1/swap`) or melt (`POST /v1/melt/{method}`) `request` Bob should immediately derive a key `k` for it. `k` should depend on the path of `request` as well as the contents of `request`'s payload.

Reviewer: ```suggestion
Upon the reception of a mint (`POST /v1/mint/{method}`), swap (`POST v1/swap`) or melt (`POST /v1/melt/{method}`) `request`, the mint derives a key `k` for it. `k` should depend on the path of `request` as well as the contents of `request`'s payload.
```

> Bob should perform a `k` based look-up `response = D[k]` and discriminate execution based  on the following check:

Reviewer: ```suggestion
The mint uses `k` to look up a `response = D[k]` and discriminates execution based on the following checks:
```

> * If `response == null`: `request` has no matching `response`. Bob should process `request` as per usual.
> * If `response != null`: `request` has a matching `response`. Bob should return the cached `response` (REPEAT).

Reviewer: ```suggestion
* If no cached `response` is found: `request` has no matching `response`. The mint processes `request` as per usual.
* If a cached `response` is found: `request` has a matching `response`. The mint returns the cached `response`.
```

I think this sounds overly technical but I've corrected your version here.

### Store

After any mint, melt and swap request is processed and the response has been deemed to be successful (`status_code == 200`), Bob should store the response in `D`.

### Expiry

Bob autonomously decides the TTL (Time To Live) for each successful response, after which time has passed it can evict the entry from `D`.

## Settings

> Support for NUT-XX **MUST** be announced as an extension to the `nuts` field of the `GetInfoResponse` object described in [NUT-6](06) and returned upon a `GET v1/info` request.

Reviewer: ```suggestion
Support for NUT-XX is announced as an extension to the `nuts` field of the `GetInfoResponse` described in [NUT-6](06).
```

> The entry must be structured as follows:

Reviewer: ```suggestion
The entry is structured as follows:
```

```json
"nuts": {
    ...,
    "xx": {
        "ttl": <int|null>,
        "cached_endpoints": [
            {
                "path": "v1/{mint|melt}/{method}",
                "method": "{POST|GET|...}",
            },
            {
                "path": "v1/swap",
                "method": "{POST|GET|...}",
            },
            ...
        ]
    }
}
```

> Where `ttl` is the amount of seconds the responses are cached for and `cached_endpoints` is a list of the routes for which caching is enabled.

Reviewer: ```suggestion
Where `ttl` is the number of seconds the responses are cached for and `cached_endpoints` is a list of the methods and paths for which caching is enabled.
```

> `path` and `method` describe respectively the cached route and its method.

Reviewer: Should be in the previous sentence.

> if `ttl` is null, the responses are to be considered cached *indefinetely*.

Reviewer: ```suggestion
If `ttl` is `null`, the responses are expected to be cached *indefinitely*.
```
